src/nn/transformer/base.rs

```diff
@@ -0,0 +1,221 @@
+use crate::prelude::*;
+use rand::Rng;
+
+/// A multi-head attention layer.
+///
+/// # Generics
+/// - `M` The embedding size of token vectors from decoder.
+/// - `N` The embedding size of token vectors from encoder.
+/// - `K` The size of the keys in self attention.
+/// - `V` The size of the values.
+/// - `H` The number of attention heads.
+///
+/// # Examples
+/// `MultiHeadAttention<8, 10, 10, 10, 2>` is an attention layer with 2 heads and 10 token, key and value dims.
+/// TODO: Doctests fail for some reason
```

Comment:
I was seeing this too on nightly, unsure why it happened and was unable to reproduce in regular tests